server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java

```diff
@@ -204,15 +215,30 @@ public void writeTo(StreamOutput streamOutput) throws IOException {
 
     static final class HandshakeResponse extends TransportResponse {
 
-        private final Version responseVersion;
-
-        HandshakeResponse(Version responseVersion) {
-            this.responseVersion = responseVersion;
+        private final Version requestVersion;
+        private final Version version;
+        private final ClusterName clusterName;
+        private final DiscoveryNode discoveryNode;
+
+        HandshakeResponse(Version requestVersion, Version responseVersion, ClusterName clusterName, DiscoveryNode discoveryNode) {
+            this.requestVersion = requestVersion;
+            this.version = responseVersion;
+            this.clusterName = clusterName;
+            this.discoveryNode = discoveryNode;
         }
 
-        private HandshakeResponse(StreamInput in) throws IOException {
+        private HandshakeResponse(Version requestVersion, StreamInput in) throws IOException {
```

Comment:
shouldn't be the stream input version be the version we use for serialization? same applies to the writeTo part below?

Reply:
The stream serialization version at this point is pre-handshake. So it is the minimum compatibility version. The idea of sending the true version in the request is so that we can serialize the response based on that. So 7.0 will send a stream version of 6.7. But in the request it will say it is actually 7.0 and the responding node can use the knowledge that the node is 7.0 to determine an appropriate response. 

I added a comment to describe this.

Reply:
well can't we call setVersion on the stream in `TcpTransport#handleRequest` in the case of a handshake then we can simply rely on it and have the invariant contained in one place? This would also prevent us from passing the version around?

Reply:
I don't think so? We have to have deserialized the the request in order to access the version we need for setting the stream. And we would still need to access the request version for serializing the response.

In this particular case we are using the immediately deserialized version to consume the rest of the response:

```
        private HandshakeResponse(Version requestVersion, StreamInput in) throws IOException {
            super.readFrom(in);
            this.requestVersion = requestVersion;
            version = Version.readVersion(in);
            // During the handshake process, nodes set their stream version to the minimum compatibility
            // version they support. When deserializing the response, we use the version the other node
            // told us that it actually is in the handshake response (`version`).
            // TODO: On backport update to 6.7
            if (requestVersion.onOrAfter(Version.V_7_0_0) && version.onOrAfter(Version.V_7_0_0)) {
                clusterName = new ClusterName(in);
                discoveryNode = new DiscoveryNode(in);
            } else {
                clusterName = null;
                discoveryNode = null;
            }
        }
```